homeassistant/components/shelly/__init__.py

```diff
@@ -140,6 +137,46 @@ def __init__(self, hass, entry, device: aioshelly.Device):
 
         self.device.subscribe_updates(self.async_set_updated_data)
 
+        self._async_remove_input_events_handler = self.async_add_listener(
+            self._async_input_events_handler
+        )
+        self._last_input_events_count = dict()
+
+    @callback
+    def _async_input_events_handler(self):
+        """Handle device input events."""
+        for block in self.device.blocks:
+            if "inputEvent" and "inputEventCnt" in block.sensor_ids:
+                channel = int(block.channel or 0) + 1
+                last_event_count = self._last_input_events_count.get(channel)
+                if last_event_count and last_event_count != block.inputEventCnt:
+                    _LOGGER.debug(
+                        "Shelly input event: device: %s, channel: %s, inputEvent: %s, inputEventCnt: %s",
+                        get_device_name(self.device),
+                        channel,
+                        block.inputEvent,
+                        block.inputEventCnt,
+                    )
+
+                    event_type = block.inputEvent
+                    if event_type != "":
+                        if event_type in INPUTS_EVENTS_DICT.keys():
+                            self.hass.bus.async_fire(
+                                "shelly.click",
+                                {
+                                    "entity_id": get_entity_name(self.device, block),
```

Comment:
Include `device_id`

Reply:
I'm not sure if using the field `entity_id` here was a good decision, this not a real sensor entity in this case since it is a momentary button and no sensor entity is created for it.
the `get_entity_name` is generating a unique name which includes the correct channel name if available in the device and the device name.
Example:
```json
{
    "event_type": "shelly.click",
    "data": {
        "entity_id": "shellyix3-A4CF12F41234 channel 3",
        "click_type": "single"
    },
    "origin": "LOCAL",
    "time_fired": "2020-11-22T12:36:26.378711+00:00",
    "context": {
        "id": "44554bc76b193866ca9df20f0c6c1a71",
        "parent_id": null,
        "user_id": null
    }
}
```
What should I use for `device_id`?
My original plan was to use device name as `device_id` ("shellyix3-A4CF12F41234" in the example above")
And add additional field `channel`.
The benefit of using `get_entity_name` is that channel names are read from the device if they exists there.

Reply:
IMO, event data should looks like this:
```text
"data": {
    "entity_id": "shellyix3-A4CF12F41234",
    "channel": 3,
    "click_type": "single"
}
```

Reply:
Device ID should be the device ID of the device in the Home Assistant device registry. Since there are no entities, you need to manually register the device https://developers.home-assistant.io/docs/device_registry_index#manual-registration

Remove `entity_id` since it's a made up value.

Reply:
Done